src/frameCompression.jl

```diff
@@ -0,0 +1,101 @@
+# Frame Compressor Codec
+# ======================
+
+struct ZstdFrameCompressor <: TranscodingStreams.Codec
+    cstream::CStream
+    level::Int
+end
+
+function Base.show(io::IO, codec::ZstdFrameCompressor)
+    print(io, summary(codec), "(level=$(codec.level))")
+end
+
+# See compressor.jl for DEFAULT_COMPRESSION_LEVEL
+
+"""
+    ZstdFrameCompressor(;level=$(DEFAULT_COMPRESSION_LEVEL))
+
+Create a new zstd compression codec using the non-streaming API.
+This is uses `ZSTD_compress2`. This compressor expects to have the
+entire input buffer to be compressed available and stores the
+decompressed length in the frame header.
+
+Arguments
+---------
+- `level`: compression level ($(MIN_CLEVEL)..$(MAX_CLEVEL))
+"""
+function ZstdFrameCompressor(;level::Integer=DEFAULT_COMPRESSION_LEVEL)
+    if !(MIN_CLEVEL ≤ level ≤ MAX_CLEVEL)
+        throw(ArgumentError("level must be within $(MIN_CLEVEL)..$(MAX_CLEVEL)"))
+    end
+    return ZstdFrameCompressor(CStream(), level)
+end
+
+const ZstdFrameCompressorStream{S} = TranscodingStream{ZstdFrameCompressor,S} where S<:IO
+
+"""
+    ZstdFrameCompressorStream(stream::IO; kwargs...)
+
+Create a new zstd compression stream (see `ZstdFrameCompressor` for `kwargs`).
+"""
+function ZstdFrameCompressorStream(stream::IO; kwargs...)
+    x, y = splitkwargs(kwargs, (:level,))
+    return TranscodingStream(ZstdFrameCompressor(;x...), stream; y...)
+end
+
+
+# Methods
+# -------
+
+function TranscodingStreams.initialize(codec::ZstdFrameCompressor)
+    code = initialize!(codec.cstream, codec.level)
+    if iserror(code)
+        throw(ZstdError(code))
+    end
+    return
+end
+
+function TranscodingStreams.finalize(codec::ZstdFrameCompressor)
+    if codec.cstream.ptr != C_NULL
+        code = free!(codec.cstream)
+        if iserror(code)
+            throw(ZstdError(code))
+        end
+        codec.cstream.ptr = C_NULL
+    end
+    return
+end
+
+function TranscodingStreams.expectedsize(codec::ZstdFrameCompressor, input::Memory)
+    code = compressed_size_bound(input.size)
+    if iserror(code)
+        throw(ZstdError(code))
+    end
+    return Int(code)
+end
+
+function TranscodingStreams.startproc(codec::ZstdFrameCompressor, mode::Symbol, error::Error)
+    code = reset!(codec.cstream, 0 #=unknown source size=#)
+    if iserror(code)
+        error[] = ZstdError(code)
+        return :error
+    end
+    return :ok
+end
+
+function TranscodingStreams.process(codec::ZstdFrameCompressor, input::Memory, output::Memory, error::Error)
+    cstream = codec.cstream
+    cstream.ibuffer.src = input.ptr
+    cstream.ibuffer.size = input.size
+    cstream.ibuffer.pos = 0
+    cstream.obuffer.dst = output.ptr
+    cstream.obuffer.size = output.size
+    cstream.obuffer.pos = 0
+    code = frameCompress!(cstream)
+    if iserror(code)
+        error[] = ZstdError(code)
+        return 0, 0, :error
+    else
+        return Int(input.size), Int(code), :end
```

Comment:
I don't think `process` can return `:end` here, generally when compressing, `:end` is only returned if `input` is empty and no extra output needs to be written for the current frame.

Reply:
We are either going to return `:end` or `:ok` here as in compression.jl:
https://github.com/JuliaIO/CodecZstd.jl/blob/4ba1be6740368ccbbf49bf2ca7dd0b440be7491c/src/compression.jl#L96

Here, unlike the streaming API, we consumed the entire input buffer and have written the entire output by invoking `ZSTD_compress2`. There is no continuation. There is no more input to process.

By the completion of `ZSTD_compress2` there are two possible outcomes. We have either successfully compressed the data into the output buffer. How else would you describe the state after `ZSTD_compress2` runs?

Reply:
The caller of `process` is allowed to break up the input data because `process` is a streaming API. Those small inputs could be appended to an internal buffer in the codec, and then only after `input.size` is set to zero, signaling there are no more bytes in the frame, `ZSTD_compress2` is called.

Reply:
What would be the best way to detect additional bytes added and throw an error?

Reply:
I think you have to assume there will be additional bytes in the frame until `process` is called with input size zero.
There should be a new `process2` similar to `ZSTD_compressStream2` with an additional `endOp` argument, to allow one-shot compression of a frame.